MoonRPGElephantLogger/LoggerManager.cpp

```diff
@@ -0,0 +1,183 @@
+                             char const* message,
+                             char const* file,
+                             int line)
+{
+    if (this->m_isRunning && this->m_currentLogLevel.load() >= static_cast<int8_t>(level))
+    {
+        this->internalQueueLog(level, output, message, file, line);
+    }
+}
+
+void LoggerManager::saveAllLogFiles() const
+{
+    using Clock = std::chrono::system_clock;
+    std::time_t startTime = Clock::to_time_t(Clock::now());
+    char timestamp[30];
+    std::strftime(timestamp, 30, "/%Y_%m_%d_%H%M%S_MoonSavedLogs/", std::localtime(&startTime));
+    auto saveFolder = std::experimental::filesystem::path(this->m_logFileSavePath + timestamp);
+    std::experimental::filesystem::create_directory(saveFolder);
+    std::experimental::filesystem::copy(this->m_logFilePath, saveFolder);
+}
+
+
+// -----------------------------------------------------------------------------
+// Internal methods
+// -----------------------------------------------------------------------------
+
+void LoggerManager::internalQueueLog(LogLevel level, LogChannel::Output output,
+                                     std::string message,
+                                     std::string file, 
+                                     const int line)
+{
+    // Message passed by copie, otherwise, local scope of variable would destroye it.
+    std::lock_guard<std::mutex> lock(m_queuesFrontAccessMutex);
+    this->m_queueLogsFront->emplace_back(level, output, std::move(message), std::move(file), line);
+}
+
+void LoggerManager::internalProcessBackQueue()
+{
+    for (LogMessage& msg : *this->m_queueLogsBack) 
+    {
+        std::string formattedMessage = msg.getFormattedMessage();
+
+        auto& coco = m_lookupChannels[msg.getLogChannel()];
+        coco->writeInChannel(formattedMessage);
+
+        if (this->m_isLogingInFile)
+        {
+            coco->writeInFile(formattedMessage);
+        }
+    }
+
+    this->m_queueLogsBack->clear();
+}
+
+void LoggerManager::internalSwapQueues()
+{
+    std::lock_guard<std::mutex> lock(m_queuesFrontAccessMutex);
+    std::swap(this->m_queueLogsFront, this->m_queueLogsBack);
+}
+
+void LoggerManager::internalStartLoggerThread()
+{
+    std::thread
+    {
+        [this]() {
+        while (this->m_isRunning)
+        {
+            std::this_thread::sleep_for(std::chrono::milliseconds{MoonRPG::LOGGER_THREAD_UPDATE_RATE_IN_MILLISECONDS});
+            this->internalProcessBackQueue();
+            this->internalSwapQueues();
+        }
+    }
+    }.detach();
+}
+
+
+// -----------------------------------------------------------------------------
+// Getter - Setters
+// -----------------------------------------------------------------------------
+
+void LoggerManager::setLogLevel(const LogLevel level)
+{
+    this->m_currentLogLevel = static_cast<int8_t>(level);
+}
+
+LogLevel LoggerManager::getLogLevel() const
+{
+    return static_cast<LogLevel>(this->m_currentLogLevel.load());
```

Comment:
Je pense que juste, return this->m_currentLogLevel fonctionne;

Reply:
ou bien plutôt return static_cast<LogLevel>(this->m_currentLogLevel); ?

Reply:
Je vais test ca, j'avais add le .load parce que bizzarement ca marchait pas sans, mais peut etre avec le static_cast. Je retest et debug ca pour essaye (Psk ouais c'est moche le load)